088-rebalance-progress-status.md

```diff
@@ -0,0 +1,334 @@
+
+In the `ConfigMap`, we will include the following fields:
+- **estimatedTimeToCompletion**: The estimated amount time it will take in minutes until partition rebalance is complete.
+- **percentageDataMovementComplete**: The percentage of the data movement of the partition rebalance that is completed e.g. values in the range [0-100]%
+- **executorState**: The “non-verbose” JSON payload from the `/kafkacruisecontrol/state?substates=executor` endpoint.
+
+The progress `ConfigMap` of an inter-broker partition rebalance would look like the following:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: my-rebalance-progress
+  …
+data:
+  estimatedTimeToCompletion: 5m [1]
+  percentageDataMovementComplete: 80% [2]
+  executorState: [3]
+  {
+   "abortingPartitions":0,
+   "averageConcurrentPartitionMovementsPerBroker":5,
+   "finishedDataMovement":0,
+   "maximumConcurrentPartitionMovementsPerBroker":5,
+   "minimumConcurrentPartitionMovementsPerBroker":5,
+   "numFinishedPartitionMovements":0,
+   "numInProgressPartitionMovements":0,
+   "numPendingPartitionMovements":20,
+   "numTotalPartitionMovements":20,
+   "state":"INTER_BROKER_REPLICA_MOVEMENT_TASK_IN_PROGRESS",
+   "totalDataToMove":0,
+   "triggeredSelfHealingTaskId":"",
+   "triggeredTaskReason":"No reason provided (Client: 172.17.0.1, Date: 2024-11-15T19:41:27Z)",
+   "triggeredUserTaskId":"0230d401-6a36-430e-9858-fac8f2edde93"
+  }
+```
+[1] The estimated time it will take the rebalance to complete based on the average rate of data transfer.
+
+[2] The percentage complete of the ongoing rebalance in the range [0-100]%
+
+[3] The “non-verbose” JSON payload from [/kafkacruisecontrol/state?substates=executor](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) endpoint.
+
+### Field values per `KafkaRebalance` State
+
+For the initial implementation, we will focus on providing the progress information in the following `KafkaRebalance` states:
+
+- `Rebalancing`
+- `Stopped`
+
+These are the states where this progress information will be able to be most accurately calculated and most useful for users. 
+We could provide this information for other states as well, such as the `ProposalReady` and `Ready` states, but it is not completely necessary, nor is it trivial. 
+Further discussion on the inclusion of the progress information for these other states can be found in the [Future Improvements](#future-improvements) section near the bottom of this proposal.
+
+All the information required for estimating the values of `estimatedTimeToCompletion` and `percentageDataMovementComplete` fields can be derived from either the Cruise Control server configurations or the [/kafkacruisecontrol/state?substates=executor](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) REST API endpoint.
```

Comment:
Updated "'or" -> "and" now that the `ProposalReady` state is being implemented as part of the proposal. This statement now makes more sense since the `ProposalReady` state estimations depend on the information from the Cruise Control server configurations, while the other states depend on the information from the `/kafkacruisecontrol/state?substates=executor`.